### ISSUE TITLE
Problem with aligning pointers on SYCL: top 32bit are zeroed out

### ISSUE THREAD

**Issue by cwpearson** [kokkos/kokkos#7116]
Attempting "normal" ways (see the top of the reproducer code) to align a scratch pointer usually yields the top 32-bits being zeroed.

Reproducer: launches 1 team. The team gets a chunk of scratch large enough to store one `Scalar` per team member. It uses one of four methods to align the raw scratch pointer, zeros the scratch, and then tries to set one element to its team rank using `Kokkos::atomic_add`. Rank 0 also prints the result of the aligned pointer function.

In the output below, we can see only method 1 does not truncate the aligned pointer. In the other cases, a strange pointer value is produced and the atomic ops fail, yielding detected errors in the result view.

```
aligned vals=0x10000430
align method 0 had 127 errors
aligned vals=0x4000000010000430
align method 1 had 0 errors
aligned vals=0x10000430
align method 2 had 127 errors
aligned vals=0x10000430
align method 3 had 127 errors
```

```c++
#include <Kokkos_Core.hpp>
#include <iostream>

// the original Kokkos Kernels implementation
template <typename T, typename InPtr>
KOKKOS_INLINE_FUNCTION T *align_kernels(InPtr p) {
  std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  // Kokkos::printf("%s:%i ptrVal=%lu\n", __FILE__, __LINE__, ptrVal);
  return reinterpret_cast<T *>((ptrVal + alignof(T) - 1) & (~(alignof(T) - 1)));
}

// an implementation that works for SYCL
template <typename T, typename InPtr>
KOKKOS_INLINE_FUNCTION T *align_naive(InPtr p) {
  std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  while (ptrVal % alignof(T)) {
    ++ptrVal;
  }
  return reinterpret_cast<T *>(ptrVal);
}

// another valid implementation
template <typename T, typename InPtr>
KOKKOS_INLINE_FUNCTION T *align_divmul(InPtr p) {
  std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  return reinterpret_cast<T *>((ptrVal + alignof(T) - 1) / alignof(T) *
                               alignof(T));
}

// the way GCC does it (roughly)
template <typename T, typename InPtr>
KOKKOS_INLINE_FUNCTION T *align_gcc(InPtr p) {
  std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  return reinterpret_cast<T *>((ptrVal - uint64_t(1) + alignof(T)) &
                               -alignof(T));
}

// Function to be executed by each team
template <typename Scalar, int TEST_FN, typename Results>
struct TeamFunction {
  TeamFunction() = default;
  TeamFunction(const Results &results) : results_(results) {}

  template <typename Team>
  KOKKOS_INLINE_FUNCTION void operator()(const Team &team) const {
    char *shmem = (char *)(team.team_shmem().get_shmem(team.team_size() * sizeof(Scalar)));
    Scalar *vals;
    if constexpr (0 == TEST_FN) {
      vals = align_kernels<Scalar>(shmem);
    } else if constexpr (1 == TEST_FN) {
      vals = align_naive<Scalar>(shmem);
    } else if constexpr (2 == TEST_FN) {
      vals = align_divmul<Scalar>(shmem);
    } else if constexpr (3 == TEST_FN) {
      vals = align_gcc<Scalar>(shmem);
    } else {
      static_assert(std::is_void_v<Results>, "Unexpected TEST_FN number");
    }

    if (team.team_rank() == 0) {
      Kokkos::printf("aligned vals=%p\n", vals);
    }

    const size_t i = team.team_rank();
    Scalar val     = team.team_rank();
    vals[i]        = 0;  // zero shared memory
    Kokkos::atomic_add(&vals[i], val);
#if 0  // debugging
    Kokkos::printf("%s:%i result(%lu) += %f yielded %f\n", __FILE__, __LINE__, i, val, vals[i]);
#endif

    results_(i) = vals[i];
  }

  size_t team_shmem_size(int team_size) const {
    return team_size * sizeof(Scalar);
  }

  Results results_;
};

// use atomic add to set result(i) = i
template <int TEST_FN, typename Device>
void test_alignPtrTo() {

  using Scalar = double;

  using MemorySpace  = typename Device::memory_space;
  using ExecSpace    = typename Device::execution_space;
  using TestView     = Kokkos::View<Scalar *, MemorySpace>;
  using TestPolicy   = Kokkos::TeamPolicy<ExecSpace>;
  int teamSize = TestPolicy(1, Kokkos::AUTO)
                           .team_size_max(TeamFunction<Scalar, TEST_FN, TestView>(),
                                          Kokkos::ParallelForTag{});
  if (teamSize > 128) {
    teamSize = 128;
  }

  ExecSpace space;

  TestView results("TestView", teamSize);
  TestPolicy policy(space, 1, teamSize);
  Kokkos::parallel_for("test alignment", policy,
                       TeamFunction<Scalar, TEST_FN, TestView>(results));

  int errs;
  Kokkos::parallel_reduce(
      Kokkos::RangePolicy<ExecSpace>(space, 0, teamSize),
      KOKKOS_LAMBDA(int i, int &lerr) { lerr += (results(i) != i); }, errs);

std::cerr << "align method " << TEST_FN << " had " << errs << " errors\n";
}




int main(int argc, char* argv[]) {
  Kokkos::initialize(argc, argv);
  
  test_alignPtrTo<0, Kokkos::DefaultExecutionSpace>();
  test_alignPtrTo<1, Kokkos::DefaultExecutionSpace>();
  test_alignPtrTo<2, Kokkos::DefaultExecutionSpace>();
  test_alignPtrTo<3, Kokkos::DefaultExecutionSpace>();
  
  Kokkos::finalize();
  return 0;
}
```

Kokkos was compiled like this on Blake at Sandia

```
salloc -p PV

source /projects/x86-64-icelake-rocky8/spack-config/blake-setup-user-module-env.sh
module purge
module load cmake intel-oneapi-compilers/2023.1.0 intel-oneapi-dpl/2022.1.0 git

export ROOT=$HOME/proj/fy24-spgemm/blake
export KOKKOS_SRC=$ROOT/kokkos
export KOKKOS_BUILD=/tmp/cwpears/kokkos-build
export KOKKOS_INSTALL=$ROOT/kokkos-install
export REPRO_SRC=$ROOT/sycl-atomics-repro
export REPRO_BUILD=/tmp/cwpears/sycl-atomics-repro-build
export ZES_ENABLE_SYSMAN=1
export ONEAPI_DEVICE_SELECTOR=ext_oneapi_level_zero:gpu

git clone --branch develop --depth 1 git@github.com:kokkos/kokkos.git $KOKKOS_SRC || true

cmake -S $KOKKOS_SRC -B $KOKKOS_BUILD \
  -DCMAKE_INSTALL_PREFIX=$KOKKOS_INSTALL \
  -DCMAKE_BUILD_TYPE=RelWithDebInfo \
  -DCMAKE_CXX_COMPILER=icpx \
  -DCMAKE_CXX_FLAGS="-fp-model=precise -Wno-deprecated-declarations -Wno-pass-failed" \
  -DBUILD_SHARED_LIBS=ON \
  -DKokkos_ENABLE_SYCL=ON \
  -DKokkos_ARCH_INTEL_PVC=ON \
  -DKokkos_ENABLE_ONEDPL=OFF \
  -DKokkos_ENABLE_TESTS=OFF
```

**Comment by crtrott** [comment]
I updated the title: clearly the problem is neither `get_shmem` nor the `atomic_add` you start with a valid pointer, and you fail `atomic_add` when you produced an invalid one.

**Comment by crtrott** [comment]
Adding:
```c++
    if (team.team_rank() == 0) {
      Kokkos::printf("shmem=%p\n", shmem);
    }
```
Before the `if constexpr` stuff makes the problem go away ...

**Comment by crtrott** [comment]
We need to report this to Intel I also see the issue with 2024.1.

**Comment by masterleinad** [comment]
A `SYCL`-only reproducer is
```C++
#include <sycl/sycl.hpp>

// the original Kokkos Kernels implementation
template <typename T, typename InPtr>
T *align_kernels(InPtr p) {
  std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  // sycl::ext::oneapi::experimental::printf("%s:%i ptrVal=%lu\n", __FILE__,
  // __LINE__, ptrVal);
  return reinterpret_cast<T *>((ptrVal + alignof(T) - 1) & (~(alignof(T) - 1)));
}

// an implementation that works for SYCL
template <typename T, typename InPtr>
T *align_naive(InPtr p) {
  std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  while (ptrVal % alignof(T)) {
    ++ptrVal;
  }
  return reinterpret_cast<T *>(ptrVal);
}

// another valid implementation
template <typename T, typename InPtr>
T *align_divmul(InPtr p) {
  std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  return reinterpret_cast<T *>((ptrVal + alignof(T) - 1) / alignof(T) *
                               alignof(T));
}

// the way GCC does it (roughly)
template <typename T, typename InPtr>
T *align_gcc(InPtr p) {
  std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  return reinterpret_cast<T *>((ptrVal - uint64_t(1) + alignof(T)) &
                               -alignof(T));
}

template <int TEST_FN>
void test_alignPtrTo() {
  using Scalar = double;

  sycl::queue q;

  q.submit([&](sycl::handler &cgh) {
     sycl::local_accessor<char, 1> team_scratch_memory_L0(
         sycl::range<1>(1 * sizeof(Scalar)), cgh);

     cgh.parallel_for(sycl::nd_range<1>{sycl::range<1>{1}, sycl::range<1>{1}},
                      [=](sycl::nd_item<1>) {
                        char *shmem =
                            team_scratch_memory_L0
                                .get_multi_ptr<sycl::access::decorated::yes>()
                                .get();

                        sycl::ext::oneapi::experimental::printf(
                            "%d unaligned vals=%p\n", TEST_FN, shmem);

                        Scalar *vals;
                        if constexpr (0 == TEST_FN) {
                          vals = align_kernels<Scalar>(shmem);
                        } else if constexpr (1 == TEST_FN) {
                          vals = align_naive<Scalar>(shmem);
                        } else if constexpr (2 == TEST_FN) {
                          vals = align_divmul<Scalar>(shmem);
                        } else if constexpr (3 == TEST_FN) {
                          vals = align_gcc<Scalar>(shmem);
                        } else {
                        }

                        sycl::ext::oneapi::experimental::printf(
                            "%d aligned vals=%p\n", TEST_FN, vals);
                      });
   }).wait_and_throw();
}

int main(int argc, char *argv[]) {
  test_alignPtrTo<0>();
  test_alignPtrTo<1>();
  test_alignPtrTo<2>();
  test_alignPtrTo<3>();
  return 0;
}
```
with output
```
0 unaligned vals=0x4000000010000000
0 aligned vals=0x10000000
1 unaligned vals=0x4000000010000000
1 aligned vals=0x4000000010000000
2 unaligned vals=0x4000000010000000
2 aligned vals=0x10000000
3 unaligned vals=0x4000000010000000
3 aligned vals=0x10000000
```

**Comment by crtrott** [comment]
Does it also go away if you add a print of shame before the constexpr section?

**Comment by masterleinad** [comment]
> Does it also go away if you add a print of shmem before the constexpr section?

No, that's the "unaligned" output.

**Comment by masterleinad** [comment]
```C++
#include <sycl/sycl.hpp>

// the original Kokkos Kernels implementation
template <typename T, typename InPtr>
T *align_kernels(InPtr p) {
  const std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  // sycl::ext::oneapi::experimental::printf("%s:%i ptrVal=%lu\n", __FILE__,
  // __LINE__, ptrVal);
  std::uintptr_t newval = (ptrVal + alignof(T) - 1) & (~(alignof(T) - 1));
  return reinterpret_cast<T*>(p)+newval-ptrVal;
}

// an implementation that works for SYCL
template <typename T, typename InPtr>
T *align_naive(InPtr p) {
  const std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  std::uintptr_t newval = ptrVal;
  while (newval % alignof(T)) {
    ++newval;
  }
  return reinterpret_cast<T*>(p)+newval-ptrVal;
}

// another valid implementation
template <typename T, typename InPtr>
T *align_divmul(InPtr p) {
  const std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  std::uintptr_t newval = ((ptrVal + alignof(T) - 1) / alignof(T) *
                               alignof(T));
  return reinterpret_cast<T*>(p)+newval-ptrVal;
}

// the way GCC does it (roughly)
template <typename T, typename InPtr>
T *align_gcc(InPtr p) {
  const std::uintptr_t ptrVal = reinterpret_cast<std::uintptr_t>(p);
  std::uintptr_t newval = (ptrVal - uint64_t(1) + alignof(T)) &
                               -alignof(T);
  return reinterpret_cast<T*>(p)+newval-ptrVal;
}

template <int TEST_FN>
void test_alignPtrTo() {
  using Scalar = double;

  sycl::queue q;

  q.submit([&](sycl::handler &cgh) {
     sycl::local_accessor<char, 1> team_scratch_memory_L0(
         sycl::range<1>(1 * sizeof(Scalar)), cgh);

     cgh.parallel_for(sycl::nd_range<1>{sycl::range<1>{1}, sycl::range<1>{1}},
                      [=](sycl::nd_item<1>) {
                        char *shmem =
                            team_scratch_memory_L0
                                .get_multi_ptr<sycl::access::decorated::yes>()
                                .get();

                        sycl::ext::oneapi::experimental::printf(
                            "%d unaligned vals=%p\n", TEST_FN, shmem);

                        Scalar *vals;
                        if constexpr (0 == TEST_FN) {
                          vals = align_kernels<Scalar>(shmem);
                        } else if constexpr (1 == TEST_FN) {
                          vals = align_naive<Scalar>(shmem);
                        } else if constexpr (2 == TEST_FN) {
                          vals = align_divmul<Scalar>(shmem);
                        } else if constexpr (3 == TEST_FN) {
                          vals = align_gcc<Scalar>(shmem);
                        } else {
                        }

                        sycl::ext::oneapi::experimental::printf(
                            "%d aligned vals=%p\n", TEST_FN, vals);
                      });
   }).wait_and_throw();
}

int main(int argc, char *argv[]) {
  test_alignPtrTo<0>();
  test_alignPtrTo<1>();
  test_alignPtrTo<2>();
  test_alignPtrTo<3>();
  return 0;
}
```
works for all versions, i.e., we are only computing the offset and applying that directly to the pointer instead of reinterpret_casting back.

**Comment by masterleinad** [comment]
This will be fixed in oneAPI 2025.0.0 with the latest compiler version on the testbeds, I'm seeing
```
> icpx -fsycl reproducer.cpp && ./a.out 
0 unaligned vals=0x4000000010000000
0 aligned vals=0x4000000010000000
1 unaligned vals=0x4000000010000000
1 aligned vals=0x4000000010000000
2 unaligned vals=0x4000000010000000
2 aligned vals=0x4000000010000000
3 unaligned vals=0x4000000010000000
3 aligned vals=0x4000000010000000
```
